src/SharedStream.mo

```diff
@@ -123,56 +125,58 @@ module {
     var concurrentChunksCounter : Nat = 0;
     public func sendChunk() : async* {
       #ok : Nat;
-      #err : { #Paused; #Busy; #SendChunkError : Text; #StreamClosed };
+      #err : ChunkError or { #Paused; #Busy; #SendChunkError : Text };
     } {
-      if (concurrentChunksCounter >= 5) {
+      if (concurrentChunksCounter >= maxConcurrentChunks_) {
         return #err(#Busy);
       };
       switch (lowestError) {
-        case (#Inf) {};
-        case (#Val _) { return #err(#Paused) };
+        case (null) {};
+        case (?le) { return #err(#Paused) };
       };
       let headId = queue.headIndex();
       streamIter.reset();
       let elements = Iter.toArray(streamIter);
-      if (elements.size() == 0) {
-        return #ok(0);
-      };
       try {
-        concurrentChunksCounter += 1;
-        let resp = await callback(streamId, elements, headId);
-        concurrentChunksCounter -= 1;
-        switch (resp) {
-          case (#err err) {
-            Timer.cancelTimer(heartbeatTimer);
-            return #err(
-              switch (err) {
-                case (#StreamClosed) #StreamClosed;
-                case (#NotRegistered) #SendChunkError("Not registered");
-              }
-            );
+        // if last call was more than 20 seconds ago, send anyway (keep-alive)
+        if (elements.size() > 0 or (Time.now() - lastChunkTimestamp) > 20_000_000_000) {
```

Comment:
The 20s constant should be configurable by constructor argument plus dynamically changeable by a setter function.

Reply:
I think it is easier to understand and less indentation if the if-statement is before and outside the try statement. Like 
```
if (elements.size() == 0 and Time.now() < lastChunkTimestamp + delta) return #ok(0);
```
Nicer to keep the try statement smaller and only enter it when we actually want to send something.

Reply:
Done